### PR TITLE
fix: G4 material recording with more than 1 tracks per event

### PR DESCRIPTION
**Pull request by Corentin-Allaire**
The MaterialSteppingAction used to assume the G4 material step were associated with tracks with increasing ID unfortunately that is not the case (in my test they where actually decreasing). This meant that material recording wouldn't work properly if more that 1 track per event was used. 

This PR fix this issue by replacing the vector material track by a map using the track id as key. This way we are sure the recording will work no matter the ordering of the tracks and material step.

This also appear to fix the issue that sometime G4 material recording would have less track than expected (I am not fully sure why)

### CODE REVIEW

File: Examples/Python/tests/test_examples.py [hunk 1]
```diff
@@ -214,12 +214,11 @@ def test_propagation(tmp_path, trk_geo, field, seq, assert_root_hash):
 @pytest.mark.skipif(not dd4hepEnabled, reason="DD4hep not set up")
 def test_material_recording(tmp_path, material_recording, assert_root_hash):
 
-    # Not quite sure why this isn't 200
     root_files = [
         (
             "geant4_material_tracks.root",
             "material-tracks",
-            198,
+            200,
```
Comment (paulgessinger) [inline Examples/Python/tests/test_examples.py:221]: Ah, so that's why! :)

**Comment by paulgessinger** [comment]
Isn't this very similar to the problem that @XiaocongAi fixed by sorting the input events? I don't know if that's the better approach, but is that not an option here?

Also, would `unordered_map` work?

**Comment by Corentin-Allaire** [comment]
> Isn't this very similar to the problem that @XiaocongAi fixed by sorting the input events? I don't know if that's the better approach, but is that not an option here?
> 
> Also, would `unordered_map` work?

Here the issue is internal to G4 (there is no input) I couldn't find where the track id are defined in G4 or where the decision is made to process one before another.

But I agree `unordered_map` should work

**Review by paulgessinger — APPROVED** [review]
LGTM!

**Pull request merged**